### ISSUE TITLE
Dashboards: "Export for sharing" sometimes leaves original data source UID in place

### ISSUE THREAD

**Issue by candlerb** [grafana/grafana#94312]
### What happened?

When I import a specific dashboard, bind it to data sources and then re-export it "for external sharing", I find that a few instances of the data source are left with a hard-coded UID instead of bring replaced with placeholders.  (Full reproducer below)

Comparing the original imported JSON with the exported-for-sharing JSON:

```
% diff -u 19727_rev2.json Incus-1728294540333.json
```

 I see plenty of expected differences:

```
-        "uid": "${DS_INCUS}"
+        "uid": "${DS_PROMETHEUS}"
```

but there are two cases where the UID of the datasource has been hard-coded into the export, which it seems would make it unsuitable for "sharing externally", whereas the imported dashboard *did* have a placeholder.

```
@@ -103,7 +103,7 @@
       "collapsed": false,
       "datasource": {
         "type": "prometheus",
-        "uid": "${DS_INCUS}"
+        "uid": "PBFA97CFB590B2093"
       },
       "gridPos": {
         "h": 1,
```

```
@@ -2396,7 +2409,7 @@
       "collapsed": true,
       "datasource": {
         "type": "prometheus",
-        "uid": "${DS_INCUS}"
+        "uid": "PBFA97CFB590B2093"
       },
       "gridPos": {
         "h": 1,
```

### What did you expect to happen?

I expected the dashboard to be exported with *all* the data sources replaced with placeholders `${DS_PROMETHEUS}` or `${DS_LOKI}`

### Did this work before?

Apparently not. The dashboard says it was created with Grafana 9.3.2, and if I repeat the process using Grafana 9.3.2 instead of 11.2.2, the problem existed back then:

```
% diff -u 19727_rev2.json Incus-1728295518592.json
...
@@ -103,7 +103,7 @@
       "collapsed": false,
       "datasource": {
         "type": "prometheus",
-        "uid": "${DS_INCUS}"
+        "uid": "PBFA97CFB590B2093"
       },
       "gridPos": {
         "h": 1,
...
@@ -2396,7 +2398,7 @@
       "collapsed": true,
       "datasource": {
         "type": "prometheus",
-        "uid": "${DS_INCUS}"
+        "uid": "PBFA97CFB590B2093"
       },
       "gridPos": {
         "h": 1,
```

And yet, the dashboard author was able to export it successfully from *their* system.

### How do we reproduce it?

1. Install grafana 11.2.2 via apt on Ubuntu 24.04
2. Create `/etc/grafana/provisioning/datasources/datasources.yaml` with
    ```
    # datasources.yaml
    apiVersion: 1
    prune: true

    datasources:
      - name: Prometheus
        type: prometheus
        access: proxy
        orgId: 1
        uid: PBFA97CFB590B2093
        url: http://localhost:9090
        isDefault: true
        version: 1
        editable: true

      - name: Loki
        type: loki
        access: proxy
        orgId: 1
        uid: P8E80F9AEF21F6940
        url: http://localhost:3100
        version: 1
        editable: true
    ```
3. Start (or restart) grafana-server
4. Login to the GUI
5. Dashboards > New > Import,  enter ID 19727 and click Load, select "Prometheus" and "Loki" as the data sources, and Import. (You'll get rev2 which is the current version). Ignore the rendering errors because prometheus and loki aren't running; the problem also occurs when they are.
6. Click on "Share", Export, enable "Export for sharing externally", save to file.
7. Compare the saved JSON with the original dashboard source: https://grafana.com/api/dashboards/19727/revisions/2/download

### Is the bug inside a dashboard panel?

Hard to tell. As far as I can see, the two affected panels are both of type "row":

```
  "panels": [
    {
      "collapsed": false,
      "datasource": {
        "type": "prometheus",
        "uid": "PBFA97CFB590B2093"
      },
      "gridPos": {
        "h": 1,
        "w": 24,
        "x": 0,
        "y": 0
      },
      "id": 319,
      "panels": [],
      "title": "Project overview",
      "type": "row"
    },
```

```
    {
      "collapsed": true,
      "datasource": {
        "type": "prometheus",
        "uid": "PBFA97CFB590B2093"
      },
      "gridPos": {
        "h": 1,
        "w": 24,
        "x": 0,
        "y": 54
      },
      "id": 6,
      "panels": [
...
      ],
      "repeat": "name",
      "title": "Instance: $name",
      "type": "row"
    },
```

However the dashboard contains three instances of "row", and the third one is not affected - although it doesn't refer to a data source.

```
    {
      "collapsed": false,
      "gridPos": {
        "h": 1,
        "w": 24,
        "x": 0,
        "y": 55
      },
      "id": 521,
      "panels": [],
      "title": "Loki logs",
      "type": "row"
    },
```

It's possible that a "row" should not have a data source assigned in the first place - in which case, perhaps it should be discarded on either import or export?

### Environment (with versions)?

Grafana: 11.2.2 or 9.3.2
OS: Ubuntu 24.04.1
Browser: Chrome/macOS

### Grafana platform?

A package manager (APT, YUM, BREW, etc.)

**Comment by evictorero** [comment]
Hi @candlerb, I'm trying to reproduce it.
I'm using the latest version of Grafana and followed your steps. I couldn't find any `uid` hardcoded. This is the file I've generated for external export [Incus-1734029671545.json](https://github.com/user-attachments/files/18116257/Incus-1734029671545.json).
Can you reproduce this issue in the latest Grafana?

**Comment by candlerb** [comment]
The problem is fixed after upgrading from grafana 11.2.2 to 11.4.0.

Comparing the exports from both versions, the spurious "datasource" entries with hard-coded UID are gone.

```
% diff -ubB ~/Downloads/Incus-1734084030908.json ~/Downloads/Incus-1734084227029.json
--- /Users/brian/Downloads/Incus-1734084030908.json	2024-12-13 10:00:30
+++ /Users/brian/Downloads/Incus-1734084227029.json	2024-12-13 10:03:47
@@ -23,7 +23,7 @@
       "type": "grafana",
       "id": "grafana",
       "name": "Grafana",
-      "version": "11.2.2"
+      "version": "11.4.0"
     },
     {
       "type": "panel",
@@ -84,7 +84,6 @@
   "description": "Overview of Incus instances",
   "editable": false,
   "fiscalYearStartMonth": 0,
-  "gnetId": 19727,
   "graphTooltip": 0,
   "id": null,
   "links": [
@@ -97,14 +96,9 @@
       "url": "https://linuxcontainers.org/incus/docs/main/"
     }
   ],
-  "liveNow": false,
   "panels": [
     {
       "collapsed": false,
-      "datasource": {
-        "type": "prometheus",
-        "uid": "000000001"
-      },
       "gridPos": {
         "h": 1,
         "w": 24,
@@ -188,7 +182,7 @@
           "sort": "none"
         }
       },
-      "pluginVersion": "7.4.1",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -293,7 +287,7 @@
           "sort": "none"
         }
       },
-      "pluginVersion": "7.4.1",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -397,7 +391,7 @@
           "sort": "none"
         }
       },
-      "pluginVersion": "7.4.1",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -480,7 +474,7 @@
         "textMode": "auto",
         "wideLayout": true
       },
-      "pluginVersion": "11.2.2",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -550,7 +544,7 @@
         "textMode": "auto",
         "wideLayout": true
       },
-      "pluginVersion": "11.2.2",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -620,7 +614,7 @@
         "textMode": "auto",
         "wideLayout": true
       },
-      "pluginVersion": "11.2.2",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -690,7 +684,7 @@
         "textMode": "auto",
         "wideLayout": true
       },
-      "pluginVersion": "11.2.2",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -781,7 +775,7 @@
           "sort": "none"
         }
       },
-      "pluginVersion": "7.4.1",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -887,7 +881,7 @@
           "sort": "none"
         }
       },
-      "pluginVersion": "7.4.1",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -993,7 +987,7 @@
           "sort": "none"
         }
       },
-      "pluginVersion": "7.4.1",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -1099,7 +1093,7 @@
           "sort": "none"
         }
       },
-      "pluginVersion": "7.4.1",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -1243,7 +1237,7 @@
           "sort": "none"
         }
       },
-      "pluginVersion": "8.1.1",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -1416,7 +1410,7 @@
           "sort": "none"
         }
       },
-      "pluginVersion": "8.1.1",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -1569,7 +1563,7 @@
           "sort": "none"
         }
       },
-      "pluginVersion": "8.1.1",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -1686,7 +1680,7 @@
           "sort": "none"
         }
       },
-      "pluginVersion": "8.1.1",
+      "pluginVersion": "11.4.0",
       "targets": [
         {
           "datasource": {
@@ -2417,10 +2411,6 @@
     },
     {
       "collapsed": true,
-      "datasource": {
-        "type": "prometheus",
-        "uid": "000000001"
-      },
       "gridPos": {
         "h": 1,
         "w": 24,
@@ -3771,8 +3761,8 @@
       "type": "logs"
     }
   ],
-  "refresh": false,
-  "schemaVersion": 39,
+  "refresh": "",
+  "schemaVersion": 40,
   "tags": [],
   "templating": {
     "list": [
@@ -3783,10 +3773,8 @@
           "uid": "${DS_PROMETHEUS}"
         },
         "definition": "label_values(incus_procs_total, job)",
-        "hide": 0,
         "includeAll": false,
         "label": "Job",
-        "multi": false,
         "name": "job",
         "options": [],
         "query": {
@@ -3795,12 +3783,8 @@
         },
         "refresh": 1,
         "regex": "",
-        "skipUrlSync": false,
         "sort": 1,
-        "tagValuesQuery": "",
-        "tagsQuery": "",
-        "type": "query",
-        "useTags": false
+        "type": "query"
       },
       {
         "current": {},
@@ -3810,10 +3794,8 @@
         },
         "definition": "label_values(incus_procs_total{job=\"$job\"}, project)",
         "description": "Name of the project",
-        "hide": 0,
         "includeAll": false,
         "label": "Project",
-        "multi": false,
         "name": "project",
         "options": [],
         "query": {
@@ -3822,15 +3804,10 @@
         },
         "refresh": 1,
         "regex": "",
-        "skipUrlSync": false,
         "sort": 1,
-        "tagValuesQuery": "",
-        "tagsQuery": "",
-        "type": "query",
-        "useTags": false
+        "type": "query"
       },
       {
-        "allValue": "",
         "current": {},
         "datasource": {
           "type": "prometheus",
@@ -3838,7 +3815,6 @@
         },
         "definition": "label_values(incus_procs_total{job=\"$job\",project=\"$project\"}, name)",
         "description": "Name of the instance",
-        "hide": 0,
         "includeAll": true,
         "label": "Instance",
         "multi": true,
@@ -3850,24 +3826,17 @@
         },
         "refresh": 1,
         "regex": "",
-        "skipUrlSync": false,
         "sort": 1,
-        "tagValuesQuery": "",
-        "tagsQuery": "",
-        "type": "query",
-        "useTags": false
+        "type": "query"
       },
       {
         "current": {
-          "selected": false,
           "text": "5",
           "value": "5"
         },
         "description": "Number of instances to include in pie charts",
-        "hide": 0,
         "includeAll": false,
         "label": "Top",
-        "multi": false,
         "name": "top",
         "options": [
           {
@@ -3902,8 +3871,6 @@
           }
         ],
         "query": "3, 4, 5, 10, 15, 20",
-        "queryValue": "",
-        "skipUrlSync": false,
         "type": "custom"
       }
     ]
```
